test/context_test.exs

```diff
@@ -84,6 +89,19 @@ defmodule ContextTest do
            """
   end
 
+  # TODO: Fix it!
+  # test "pass context to child component with @inner_content" do
+  #   code = """
+  #   <OuterUsingInnerContent>
+  #     <Inner/>
+  #   </OuterUsingInnerContent>
+  #   """
+
+  #   assert render_live(code) =~ """
+  #          <span id="field">field from OuterUsingInnerContent</span>\
+  #          """
+  # end
```

Comment:
@lnr0626 I'm not sure why this test fails yet. I suspect the ContentHandler is broken for this case.

Reply:
I think the issue is that context sets are no longer added to the overall components assigns, and so the ContentHandler doesn't have any way of knowing the updated value of `__context__`. in the following code snippet args[:__context__] would be nil, and so it would fall back to the components assign.__context__:

https://github.com/msaraiva/surface/blob/dc2dc34c7452f81475fc8938de7bbbd712c4f0ef/lib/surface/content_handler.ex#L88-L102

The only way to make this work correctly (that I can think of) is to require passing `__context__: @__context__` when using `inner_content`. We could attempt to inject this automatically (as we already know where the `inner_content` calls are), but that feels like it would be error prone and complex. We could also just document this and strongly recommend always using `<slot />`.

Reply:
> We could also just document this and strongly recommend always using `<slot />`

We could but, in this case, we need to find a way to use `<slot/>` when working with multiple named slots. Currently, the only way to render those slots is traversing them and calling `inner_content` explicitly:

```jsx
<tr :for={{ item <- @items }}>
  <td :for={{ col <- @cols }}>
    {{ col.inner_content.(item: item, info: info) }}
  </td>
</tr>
```

We need to accept something like:

```jsx
<tr :for={{ item <- @items }}>
  <td :for={{ col <- @cols }}>
    <slot instance={{ col }} props={{ item: item, info: info }}/>
  </td>
</tr>
```

There's no `instance` attribute defined in the HTML's `<slot>` specification so maybe we should add it as a directive `:instance` or maybe even use the existing `name`:

```jsx
<tr :for={{ item <- @items }}>
  <td :for={{ col <- @cols }}>
    <slot name={{ col }} props={{ item: item, info: info }}/>
  </td>
</tr>
```

Not sure.

Reply:
overloading `name` seems like a reasonable approach - though it'd be hard to validate at compile time that the slot was actually defined

Reply:
I agree - using `slotable` here would feel a bit weird, but I can't quite verbalize why. I think it might have something to do with "slotable" being an adjective - it feels a bit more natural to say "the slotable entry/item" or even just "the entry/item" than "the slotable".

I prefer either `entry` or `item`; but don't have a strong preference between the two, so let's do `:entry`

For directive vs. attribute - are you trying to avoid defining non-standard attributes?

Reply:
> For directive vs. attribute - are you trying to avoid defining non-standard attributes?

Yes. One thing I want to avoid is confusion when the user needs to remember if something is a `property`, `attribute` or `directive`. Today we have a few very basic rules:

  1. Components have properties which have flexible semantics, i.e. the component's author defines its runtime behaviour.
  2. HTML tags have attributes which are not flexible because we don't own them. We can still provide some syntact surgar but never change its semantics.
  5. Any aditional behaviour that can't be provided using a property or is not already provided by an attribute has to be a directive.

The only attributes we explicitly add behaviour to are `slot` and `name` of `<template>` and `<slot>`, respectvely. However, we only do that because that's exactly the expected behaviour as defined in their specs. We don't change the semantics of `slot` nor `name`.

In order to keep those rules unbroken for `<slot>`, we have two options:

  1. Add the `:entry` directive
  2. Use a component `<Slot>`(or macro component `<#Slot>`) instead of the standard HTML tag `<slot>`

Reply:
I wonder if we could simplify the whole thing and just create a macro for it. Something like:

In `Surface.ex`:

```elixir
  defmacro render_slot(slot, props \\ []) do
    quote do
      unquote(slot).inner_content.([__context__: @__context__] ++ unquote(props))
    end
  end
```

Then using it:

```jsx
<td :for={{ col <- @cols }}>
  {{ render_slot(col, item: item, info: info) }}
</td>
```

The macro should just replace the call with:

```elixir
col.inner_content.([__context__: @__context__] ++ [item: item, info: info])
```

I tried the code above but it didn't work. I got this:

```
warning: undefined module attribute @__context__, please remove access to @__context__ or explicitly set it before access
```

The generated code is:

```elixir
for(col <- @cols) do
  arg0 = Phoenix.LiveView.Engine.to_safe(render_slot(col, [item: item, info: info]))
  {:safe, ["<span>\n    ", arg0, "\n  </span>"]}
end
```

Apparently the macro didn't expand correctly.

Reply:
That wouldn't let us handle template `:let` the way that we do know; however I don't know why that didn't expand correctly.

Reply:
Hm. I would have expected that call to the macro to be expanded

Reply:
ohh, it's because the macro is expanded after it's passed through the phoenix live view engine - which is what handles changing the `@name` syntax into `assigns.name` and setting up change tracking based on that value.

Reply:
An idea - we could potentially use the `Context` component to update the `inner_content` assigns to add in `__context__` and pass all other arguments through - it would likely require compiler support to pass the slot assigns and the `@inner_content` assign, and I'm unsure of the complexity; but it would allow continued use of `@inner_content` and `col.inner_content`.... I'm also unsure of the implications for change tracking (though assuming that was an issue, we could potentially track usage of `@inner_content` and `@slot_name` assigns and only do that logic if either of those were referenced)

We could test this out by just passing those assigns in manually (i.e. `inner_content={{ @inner_content }}`), and basically rebind them to be: `fn args -> original_func.([__context__: updated_ctx] ++ args)`. for the slots, we'd have to map over them and update the inner contents within to something similar.

Reply:
We'd only have to do that within `<Context>` that used set

Reply:
That would also allow us to (slightly) simplify the conversion code for `<slot />` as we would no longer have to inject `@__context__` (in essence we could delete https://github.com/msaraiva/surface/blob/master/lib/surface/compiler/eex_engine.ex#L133-L136)

It also would make `<slot :props={{ props }} />` identical to `slot_name.inner_content(props)`, and 
```
<slot :props={{ props }}>
  <!-- default content -->
</slot>
```
identical to 
```
if slot_name do
  slot_name.inner_content(props)
else
   "<!-- default content -->"
end
```

Reply:
Ok. I implemented the `<slot :entry={{...}}/>` and play with it for a while and the result is that I didn't like it :(

Its use is not intuitive and It makes the code much more complex than it should.

> it would likely require compiler support to pass the slot assigns and the @inner_content assign

If you're talking about changing `inner_content` calls at compile-time then I think it's probably the way to go. I might be wrong but I have a feeling that if we to identity `inner_content` calls at compile-time and change them accordingly, the whole thing will become much simpler. Maybe even the whole `ContentHandler` could disappear. 

I'm back to the drawing board.